### PR TITLE
make use of ringpop LookN method

### PR DESCRIPTION
**Pull request by ast2023**
**What changed?**
To find number of per namespace workers we now use ringpop LookupN method instead of sequentially calling Lookup. The lookup key also changed: worker id was dropped from the hashed string.

Added util.SliceMap function and renamed util.SliceReduce to util.SliceFold

**Why?**
more evenly distribute work across ring

**How did you test it?**
local testing only

**Potential risks**
not sure

**Is hotfix candidate?**
no

### CODE REVIEW

File: service/worker/pernamespaceworker.go [hunk 1]
```diff
@@ -262,21 +262,19 @@ func (wm *perNamespaceWorkerManager) removeWorker(ns *namespace.Namespace) {
...
-	// multiple lookups might land on the same node. To compensate, we increase the number of
-	// pollers in that case, but it would be better to try to spread them across our nodes.
-	// TODO: implement this properly using LookupN in serviceResolver
-	multiplicity := 0
-	for i := 0; i < totalWorkers; i++ {
-		key := fmt.Sprintf("%s/%d", ns.ID().String(), i)
-		target, err := wm.serviceResolver.Lookup(key)
-		if err != nil {
-			return 0, 0, err
+	key := ns.ID().String()
```
Comment (ast2023) [inline service/worker/pernamespaceworker.go:265]: @dnr I am not sure my choice of the key is correct. Please advise.

File: common/membership/interfaces.go [hunk 2]
```diff
@@ -82,6 +82,9 @@ type (
 	ServiceResolver interface {
 		// Lookup looks up the host that currently owns the resource identified by the given key.
 		Lookup(key string) (HostInfo, error)
+		// LookupN looks n hosts that owns the resource identified by the given key, if n greater than total number
+		// of hosts total number of hosts will be returned
```
Comment (dnr) [inline common/membership/interfaces.go:86]: I wish ringpop could return exactly n results here (with duplication if necessary)... I wonder if it's worth implementing that extension ourselves

File: service/worker/pernamespaceworker.go [hunk 3]
```diff
@@ -262,21 +262,18 @@ func (wm *perNamespaceWorkerManager) removeWorker(ns *namespace.Namespace) {
...
+	targets := wm.serviceResolver.LookupN(key, totalWorkers)
+	if len(targets) == 0 {
+		return 0, 0, membership.ErrInsufficientHosts
+	}
+	IsLocal := func(info membership.HostInfo) bool { return info.Identity() == wm.self.Identity() }
+	multiplicity := util.FoldSlice(targets, 0, func(acc int, t membership.HostInfo) int {
```
Comment (dnr) [inline service/worker/pernamespaceworker.go:271]: the larger concern is: with the semantics of LookupN provided by ringpop, we'll never have multiplicity > 1 here. I'd like to have that ability so that there's only one tuning knob required to provide more pernsworker resources. (well okay, there's still a knob for number of nodes.)

I think we should handle the case where n > nodes either here or in serviceResolver

File: service/worker/pernamespaceworker_test.go [hunk 4]
```diff
@@ -202,9 +202,11 @@ func (s *perNsWorkerManagerSuite) TestMultiplicity() {
 		Enabled: false,
 	}).AnyTimes()
 
-	s.serviceResolver.EXPECT().Lookup("ns3/0").Return(membership.NewHostInfoFromAddress("self"), nil)
-	s.serviceResolver.EXPECT().Lookup("ns3/1").Return(membership.NewHostInfoFromAddress("other"), nil)
-	s.serviceResolver.EXPECT().Lookup("ns3/2").Return(membership.NewHostInfoFromAddress("self"), nil)
+	s.serviceResolver.EXPECT().LookupN("ns3", 3).Return([]membership.HostInfo{
```
Comment (dnr) [inline service/worker/pernamespaceworker_test.go:205]: this is not a possible return value from serviceResolver, as currently implemented

File: common/membership/interfaces.go (hunk 2 above)
Comment (ast2023) [reply common/membership/interfaces.go:86]: @dnr thank you for the review. What you suggest is to return the same HostInfo more then once if n is greater then len of []HostInfo returned by ringpop? Do I understand it right. Something like:
```
result := ringpop.LookN(key, n)
repeat := ceiling(n / result)
result := RepeatSlice(result, repeat)[:n]
```

File: service/worker/pernamespaceworker_test.go (hunk 4 above)
Comment (ast2023) [reply service/worker/pernamespaceworker_test.go:205]: @dnr could you elaborate?

Comment (dnr) [reply service/worker/pernamespaceworker_test.go:205]: I mean in the current state of the PR, serviceResolver never returns duplicates. So the test is fine for what it is, but the multiplicity stuff won't actually happen in practice and we want it to. That's the danger with using mocks of course. If we wanted to do better: I don't think we want to fire up ringpop in a unit test, but maybe in a functional test.

File: common/membership/interfaces.go (hunk 2 above)
Comment (ast2023) [reply common/membership/interfaces.go:86]: done

**Comment by ast2023** [comment]
Hey @dnr, I've uploaded a new version. I addressed most of the suggestions you pointed out. Additionally, I did some refactoring to ensure that the code better aligns with the information you provided. Let me know your thoughts!

File: service/worker/pernamespaceworker.go [hunk 5]
```diff
@@ -260,24 +275,43 @@ func (wm *perNamespaceWorkerManager) removeWorker(ns *namespace.Namespace) {
...
+}
+
+func (wm *perNamespaceWorkerManager) getLocallyDesiredWorkersCount(ns *namespace.Namespace, desiredNumberOfWorkers uint16) (uint16, error) {
+	key := ns.ID().String()
+	hostsInNamespace := wm.serviceResolver.LookupN(key, desiredNumberOfWorkers)
+	hostsCount := uint16(len(hostsInNamespace))
+	if hostsCount == 0 {
+		return 0, membership.ErrInsufficientHosts
+	}
+	maxWorkersPerHost := int(math.Ceil(float64(desiredNumberOfWorkers) / float64(hostsCount)))
```
Comment (dnr) [inline service/worker/pernamespaceworker.go:309]: integer math seems simpler? (and you can overshoot)
```suggestion
	maxWorkersPerHost := desiredNumberOfWorkers/hostsCount + 1
```

File: service/worker/pernamespaceworker.go [hunk 6]
```diff
@@ -262,21 +262,18 @@ func (wm *perNamespaceWorkerManager) removeWorker(ns *namespace.Namespace) {
...
-		key := fmt.Sprintf("%s/%d", ns.ID().String(), i)
-		target, err := wm.serviceResolver.Lookup(key)
-		if err != nil {
-			return 0, 0, err
+	key := ns.ID().String()
+	targets := wm.serviceResolver.LookupN(key, totalWorkers)
+	if len(targets) == 0 {
+		return 0, 0, membership.ErrInsufficientHosts
+	}
+	IsLocal := func(info membership.HostInfo) bool { return info.Identity() == wm.self.Identity() }
```
[the remark this answers is not in the document]
Comment (dnr) [reply service/worker/pernamespaceworker.go:312]: okay but name it with an initial lowercase since it's not exported

File: service/worker/pernamespaceworker_test.go [hunk 7]
```diff
@@ -202,9 +202,11 @@ func (s *perNsWorkerManagerSuite) TestMultiplicity() {
 		Enabled: false,
 	}).AnyTimes()
 
-	s.serviceResolver.EXPECT().Lookup("ns3/0").Return(membership.NewHostInfoFromAddress("self"), nil)
-	s.serviceResolver.EXPECT().Lookup("ns3/1").Return(membership.NewHostInfoFromAddress("other"), nil)
-	s.serviceResolver.EXPECT().Lookup("ns3/2").Return(membership.NewHostInfoFromAddress("self"), nil)
+	s.serviceResolver.EXPECT().LookupN("ns3", uint16(3)).Return([]membership.HostInfo{
+		membership.NewHostInfoFromAddress("self"),
+		membership.NewHostInfoFromAddress("other"),
+		membership.NewHostInfoFromAddress("self"),
```
Comment (dnr) [inline service/worker/pernamespaceworker_test.go:208]: I would like to see the mocks in the test not return duplicates, since the real version doesn't, and then test that it starts the right number of workers

File: common/membership/ringpop/service_resolver.go [hunk 8]
```diff
@@ -154,6 +155,15 @@ func (r *serviceResolver) Lookup(key string) (membership.HostInfo, error) {
 	return newHostInfo(addr, r.getLabelsMap()), nil
 }
 
+func (r *serviceResolver) LookupN(key string, n uint16) []membership.HostInfo {
+	addresses := r.ring().LookupN(key, int(n))
+	if len(addresses) == 0 {
+		r.RequestRefresh()
+		return nil
+	}
+	return util.MapSlice(addresses, membership.NewHostInfoFromAddress)
```
[the remark this answers is not in the document]
Comment (ast2023) [reply common/membership/ringpop/service_resolver.go:164]: I am confused. Per my understanding  **LookupN** returns slice of _membership.HostInfo_ while **Lookup** returns _membership.HostInfo_. 
**membersip.NewHostInfoFromAddress** is an initialization function which produces _membership.HostInfo_ from address. 
Could you please provide clarification regarding the specific aspect of concern or ambiguity you have in mind?

File: common/util/util.go [hunk 9]
```diff
@@ -128,15 +140,29 @@ func FilterSlice[T any](in []T, predicate func(T) bool) []T {
...
+func RepeatSlice[T any](xs []T, n int) []T {
+	if xs == nil {
+		return xs
+	}
+	if n <= 0 {
+		return []T{}
+	}
+	ys := make([]T, n*len(xs))
+	for i := 0; i < n; i++ {
+		copy(ys[i*len(xs):(i+1)*len(xs)], xs)
```
[the remark this answers is not in the document]
Comment (ast2023) [reply common/util/util.go:160]: 👍 thank you!

File: service/worker/pernamespaceworker_test.go [hunk 10]
```diff
@@ -192,8 +192,23 @@ func (s *perNsWorkerManagerSuite) TestEnabled() {
...
+	desiredWorkersNumber := 6
+	machinesOwningNamespace := []membership.HostInfo{
+		membership.NewHostInfoFromAddress("other-1"),
+		membership.NewHostInfoFromAddress("other-2"),
+		membership.NewHostInfoFromAddress("self"),
+	}
+	expectedMultiplicity := 2
+
+	ns := testns("ns3", enumspb.NAMESPACE_STATE_REGISTERED)
+	s.manager.config.PerNamespaceWorkerCount = func(_ string) int { return desiredWorkersNumber }
```
Comment (dnr) [inline service/worker/pernamespaceworker_test.go:211]: don't override this here, these are configured here: https://github.com/temporalio/temporal/blob/main/service/worker/pernamespaceworker_test.go#L87-L89
(or you could go the other way and move all the Config overrides to local to each test. I guess that's nicer in some ways.)

File: common/dynamicconfig/collection.go [hunk 11]
```diff
@@ -92,6 +92,9 @@ const (
 	errCountLogThreshold = 1000
 )
 
+// ErrInvalidConfiguration indicates that the value provided by dynamic config is not legal
+var ErrInvalidConfiguration = errors.New("invalid dynamic configuration")
```
[the remark this answers is not in the document]
Comment (ast2023) [reply common/dynamicconfig/collection.go:96]: I've given this some thought. The ErrInvalidConfiguration error generated by getConfiguredWorkersCountFor(namespace) wrapper function. The functions only responsibility is to validate the value retrieved from dynamic config—whether it's from the configuration file or defaults if the value can't be read.

To keep things simple and easy to handle down the road, it makes sense to put all the rules for checking in one place. The natural spot for these rules is in the config reader. This way, we get to choose: do we want to give a thumbs-up during system startup if the config isn't playing by the rules, or just roll with the default values?

Considering this rationale, the checks outlined in getConfiguredWorkersCountFor need to be executed as part of the matchAndConvert process within the dynamicconfig package. Consequently, I've designated the error as a generic error within the dynamicconfig package.

File: common/membership/interfaces.go [hunk 12]
```diff
@@ -82,6 +82,9 @@ type (
...
+		LookupN(key string, n uint16) []HostInfo
```
[the remark this answers is not in the document]
Comment (ast2023) [reply common/membership/interfaces.go]: I switched from uint16 to uint, not restricting the size. The choice of uint is deliberate; this interface not only averts incorrect usage but also eliminates the need for runtime checks. My inclination towards this approach is influenced by a background in functional languages, where a robust type system preemptively addresses many issues. If this doesn't align with Go's conventions, I'm open to switching it to int.

File: common/util/util.go [hunk 13]
```diff
@@ -128,15 +140,29 @@ func FilterSlice[T any](in []T, predicate func(T) bool) []T {
...
 	return acc
 }
 
+func RepeatSlice[T any](xs []T, n int) []T {
+	if xs == nil {
+		return xs
+	}
+	if n <= 0 {
+		return []T{}
+	}
```
[the remark this answers is not in the document]
Comment (ast2023) [reply common/util/util.go]: Thank you, done.

File: service/worker/pernamespaceworker.go [hunk 14]
```diff
@@ -260,24 +275,43 @@ func (wm *perNamespaceWorkerManager) removeWorker(ns *namespace.Namespace) {
...
+		err := fmt.Errorf("%w namespace %s, workers count %d", dynamicconfig.ErrInvalidConfiguration, ns.Name(), totalWorkers)
+		return 0, err
 	}
-	return multiplicity, totalWorkers, nil
+	return uint16(totalWorkers), nil
+}
+
+func (wm *perNamespaceWorkerManager) getLocallyDesiredWorkersCount(ns *namespace.Namespace, desiredNumberOfWorkers uint16) (uint16, error) {
+	key := ns.ID().String()
+	hostsInNamespace := wm.serviceResolver.LookupN(key, desiredNumberOfWorkers)
```
[the remark this answers is not in the document]
Comment (ast2023) [reply service/worker/pernamespaceworker.go]: Changed name to `availableHosts`. What do you think?

File: service/worker/pernamespaceworker.go (hunk 5 above)
Comment (ast2023) [reply service/worker/pernamespaceworker.go]: 👍 Thank you!

File: common/dynamicconfig/collection.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (dnr) [reply common/dynamicconfig/collection.go:96]: More validation within dynamicconfig may be a good thing, but that's not happening in this PR. So this error shouldn't be here. If this package does more validation, then it can have more errors.

Note that dynamic config usage is pretty free-form, so adding validation won't be easy. Clients can ask for any values at any time, and each request might come with a different default value (and validation options if added). Keys are not all known statically. Plus, we have to keep the interface backwards-compatible to avoid breaking code that embeds the server.

(Also I think you mean heads-up, not thumbs-up :) )

File: common/membership/interfaces.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (dnr) [reply common/membership/interfaces.go]: It does not align with Go's conventions.

Or any C-like languages tbh. Many many sources recommend against using unsigned integers for representing quantities. A few from a quick search:

https://www.learncpp.com/cpp-tutorial/unsigned-integers-and-why-to-avoid-them/
https://hamstergene.github.io/posts/2021-10-30-do-not-use-unsigned-for-nonnegativity/
https://jacobegner.blogspot.com/2019/11/unsigned-integers-are-dangerous.html
From a Go developer:
https://www.github.com/golang/go/issues/27460#issuecomment-418203686

Basically, it makes it too easy to create overflow bugs, and has no real benefits. For LookupN, the semantics are really that n should be >= 1. Using an unsigned int only excludes some, but not all, invalid values.

[the remark this answers is not in the document]
Comment (ast2023) [reply common/membership/interfaces.go]: Thanks. I looked for natural number, but Go does not have one and there seems to be no easy way to create one. So, I used uint. I will switch to int.

File: common/dynamicconfig/collection.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (dnr) [reply common/dynamicconfig/collection.go:96]: Please remove this error

File: common/util/util_test.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package util
+
+import "testing"
+
+func TestRepeatSlice(t *testing.T) {
+	t.Run("when input slice is nil should return nil", func(t *testing.T) {
```
Comment (dnr) [inline common/util/util_test.go:30]: This isn't a blocker, but I'll just note that these tests are written in a fairly different style from the rest of the unit tests in the repo, which generally use helpers from `github.com/stretchr/testify/require` or `assert` and `suite` from that project. I'd prefer if these matched the prevailing style.

File: service/worker/pernamespaceworker.go [hunk 16]
```diff
@@ -260,24 +267,43 @@ func (wm *perNamespaceWorkerManager) removeWorker(ns *namespace.Namespace) {
...
+
+func (wm *perNamespaceWorkerManager) getLocallyDesiredWorkersCount(ns *namespace.Namespace, desiredNumberOfWorkers int) (int, error) {
+	key := ns.ID().String()
+	availableHosts := wm.serviceResolver.LookupN(key, desiredNumberOfWorkers)
+	hostsCount := len(availableHosts)
+	if hostsCount == 0 {
+		return 0, membership.ErrInsufficientHosts
+	}
+	maxWorkersPerHost := desiredNumberOfWorkers/hostsCount + 1
+	desiredDistribution := util.RepeatSlice(availableHosts, int(maxWorkersPerHost))[:desiredNumberOfWorkers]
```
Comment (dnr) [inline service/worker/pernamespaceworker.go:302]: do you need the `int()`? maxWorkersPerHost should be an `int` already, right?

File: service/worker/pernamespaceworker.go [hunk 17]
```diff
@@ -465,19 +490,19 @@ func (w *perNamespaceWorker) startWorker(
...
+	sdkoptions.MaxConcurrentActivityExecutionSize *= int(allocation.Local)
 	sdkoptions.OnFatalError = w.onFatalError
 
 	// this should not block because the client already has server capabilities
 	worker := w.wm.sdkClientFactory.NewWorker(client, primitives.PerNSWorkerTaskQueue, sdkoptions)
 	details := workercommon.RegistrationDetails{
-		TotalWorkers: totalWorkers,
-		Multiplicity: multiplicity,
+		TotalWorkers: int(allocation.Total),
+		Multiplicity: int(allocation.Local),
```
Comment (dnr) [inline service/worker/pernamespaceworker.go:505]: no need for any of these casts

File: service/worker/pernamespaceworker_test.go [hunk 18]
```diff
@@ -192,8 +192,22 @@ func (s *perNsWorkerManagerSuite) TestEnabled() {
...
+	    fair distribution across all the machines owning the namespace
+*/
 func (s *perNsWorkerManagerSuite) TestMultiplicity() {
-	ns := testns("ns3", enumspb.NAMESPACE_STATE_REGISTERED) // three workers
+	machinesOwningNamespace := []membership.HostInfo{
+		membership.NewHostInfoFromAddress("other-1"),
+		membership.NewHostInfoFromAddress("other-2"),
+		membership.NewHostInfoFromAddress("self"),
+	}
+	desiredWorkersNumber := 6
```
Comment (dnr) [inline service/worker/pernamespaceworker_test.go:207]: does this have to match the `6` on line 88? if so, either use a single shared constant, or add a comment explaining that they're linked

File: service/worker/pernamespaceworker_test.go [hunk 19]
```diff
@@ -288,9 +314,11 @@ func (s *perNsWorkerManagerSuite) TestTwoNamespacesTwoComponents() {
 			return &workercommon.PerNSDedicatedWorkerOptions{Enabled: ns.Name().String() == "ns1"}
 		}).AnyTimes()
 
-	s.serviceResolver.EXPECT().Lookup("ns1/0").Return(membership.NewHostInfoFromAddress("self"), nil)
-	s.serviceResolver.EXPECT().Lookup("ns2/0").Return(membership.NewHostInfoFromAddress("self"), nil)
-	s.serviceResolver.EXPECT().Lookup("ns2/1").Return(membership.NewHostInfoFromAddress("self"), nil)
+	s.serviceResolver.EXPECT().LookupN("ns1", 1).Return([]membership.HostInfo{membership.NewHostInfoFromAddress("self")})
+	s.serviceResolver.EXPECT().LookupN("ns2", 2).Return([]membership.HostInfo{
```
Comment (dnr) [inline service/worker/pernamespaceworker_test.go:318]: same for this one, LookupN never returns duplicates itself

File: common/util/util_test.go (hunk 15 above)
Comment (ast2023) [reply common/util/util_test.go:34]: switched to assert

**Review by dnr — APPROVED** [review]
Please fix the typo in the PR title
